[Unity] Merge with recent unity changes

jwfromm opened the pull request:
This PR is the first merge between our repo and upstream tvm/unity rather than rebasing. The merge went quite smoothly locally. Please give the changes a look and make sure there are no surprises.

testing - c7200b7

YuchenJin commented on src/relax/ir/expr_functor.cc:
```diff
@@ -354,9 +354,17 @@ void ExprMutator::VisitBinding_(const VarBindingNode* binding) {
+  };
+
   if (new_var.same_as(binding->var) && new_value.same_as(binding->value)) {
     // no-op if there is no change
-    builder_->Emit(GetRef<VarBinding>(binding));
+    emit(GetRef<VarBinding>(binding));
     return;
   }
```
This fast path may cause problems when the `binding->value`'s type/shape fields are filled while the `binding->var`'s type/shape are unfilled (which is the default case when user does not annotate the shape and type of the Var in TVMScript). If we expect the lhs var's type/shape also to be filled, we should remove this fast path.

altanh replied:
commented out for now, we'll try to address all these quirks by moving more of the boilerplate fast path stuff to the normalizer/builder

YuchenJin replied:
Sounds good! :)

YuchenJin reviewed (approved):
LGTM! Thanks @altanh for the effort!

ZihengJiang commented on src/relax/ir/expr_functor.cc:
```diff
@@ -354,11 +354,20 @@ void ExprMutator::VisitBinding_(const VarBindingNode* binding) {
+    }
+  };
+
+  // FIXME(@altanh): try to clean up all the fast paths and ty/shape infer, it's getting unwieldy
```
remove?

altanh replied:
I want to leave this here since we'll revisit it later, since not 100% sure the current fix is a proper fix

ZihengJiang commented on python/tvm/relax/transform/transform.py:
```diff
@@ -76,11 +76,15 @@ def VMShapeLower() -> tvm.transform.Pass:
     return _ffi_api.ToANF()
+
+
+def ResolveGlobals() -> tvm.ir.transform.Pass:
```
comment

altanh replied:
good catch, thx

ZihengJiang commented on tests/python/relax/test_transform.py:
```diff
@@ -211,30 +212,30 @@ def foo(x: Tensor[_, "float32"]) -> Shape:
+            gv0 = R.call_dps((m, k), tir_matmul, (x, w))
+            return gv0
+
+    mod = InputModule
```
Can we change the syntax to `InputModule()`?

altanh replied:
this would change the existing behavior used by TVMScript so I avoided it for now, but if the TVM script maintainers agree then we can make the change

ZihengJiang replied:
I see @junrushao1994

YuchenJin replied:
Also cc @Hzfengsy

ZihengJiang commented:
LGTM! Feel free to merge.

psrivas2 committed:
[Unity][Pass] Enhance constant folding to fold relax ops by evaluating them. (#14146)

* [Unity][Pass] Enhance constant folding to fold relax ops
by evaluating them.

This uses the registered legalization function attached to
the op to lower it to call_tir and uses the existing call_tir
folding mechanism to fold it.

This kind of op folding is only allowed within dataflow block
as ops could have side-effects.

Limitations:
* This currently does not support folding ops
that could lower to multiple call_tir bindings.
* Folding by evaluating ops is not always beneficial.
We need a heuristic to check if it is useful. This is
not implemented yet and folding is always allowed
by evaluating expressions.

* fix ci error

* fix doc

* fix bug

slyubomirsky committed:
[Unity][Debugging] AST printer (#14152)

This PR transfers over the AST printer from tlc-pack/relax. The AST printer is a debugging tool that prints out a Relax AST in a precise and human-readable format, which can be helpful for debugging the parser or various passes.

Co-authored-by: Yuchen Jin <yuchenj@cs.washington.edu>
Co-authored-by: Lesheng Jin <34279105+LeshengJin@users.noreply.github.com>
Co-authored-by: Siyuan Feng <Hzfengsy@sjtu.edu.cn>
Co-authored-by: Ruihang Lai <ruihangl@cd.cmu.edu>
Co-authored-by: Tianqi Chen <tqchen@users.noreply.github.com>

Commit:
[Unity][Pass] Support Symbolic Shape Deduction during BindParam (#14154)

`BindParam` replace function params to constant nodes. However, it will
drop the shape information of the params, considering the following case:

```python
@R.function
def main(
    x: R.Tensor(("batch", "m"), dtype="float32"),
    w0: R.Tensor(("n", "m"), dtype="float32"),
    b0: R.Tensor(("n",), dtype="float32"),
    w1: R.Tensor(("k", "n"), dtype="float32"),
    b1: R.Tensor(("k",), dtype="float32"),
) -> R.Tensor(("batch", "k"), dtype="float32"):
    batch = T.Var("batch", "int64")
    k = T.Var("k", "int64")
    m = T.Var("m", "int64")
    n = T.Var("n", "int64")
    with R.dataflow():
        lv0 = R.call_tir("linear0", (x, w0, b0), out_sinfo=R.Tensor((batch, n), dtype="float32"))
        out = R.call_tir("linear1", (lv0, w1, b1), out_sinfo=R.Tensor((batch, k), dtype="float32"))
        R.output(out)
    return out
```

The current pass will simply drop the symbolic var `n`, `k` and cause
undefined vars during build as
```python
@R.function
def main(x: R.Tensor((1, "m"), dtype="float32")) -> R.Tensor(dtype="float32", ndim=2):
    m = T.Var("m", "int64")
    n = T.Var("n", "int64")
    k = T.Var("k", "int64")
    with R.dataflow():
        lv0 = R.call_tir("linear0", (x, metadata["relax.expr.Constant"][0], metadata["relax.expr.Constant"][1]), out_sinfo=R.Tensor((1, n), dtype="float32"))
        out = R.call_tir("linear1", (lv0, metadata["relax.expr.Constant"][2], metadata["relax.expr.Constant"][3]), out_sinfo=R.Tensor((1, k), dtype="float32"))
        R.output(out)
    return out
```

This PR updates the pass to bind the symbolic shape during binding.

Commit:
[Unity][Analysis] Checking function return struct info in well-formed check (#14155)

The current well-formed misses the check of function return struct info,
which may mistakenly pass the check if there are undefined vars in the
function return struct info.

masahi committed:
[Unity][BYOC] Use Relax legalize + CPU build for reference in tests (#14162)

* clean dnnl test

* clean trt test

* clean cutlass test

* fix gelu legalize for fp16

* use memoize in dnnl and trt tests

vinx13 committed:
[Unity] Add bind_constants option to FuseOpsByPattern (#14151)

* [Unity] Add lift_constatns option to FuseOpsByPattern

* lift_constants -> bind_constants

slyubomirsky committed:
[Unity][Analysis] Analysis for detecting recursion in Relax (#14149)

* DFS based attempt to detect mutual recursion

* Use Johnson's circuit-detecting algorithm instead

* Fix control flow test

* Detect all recursion anyway

* Add new test cases for simple recursion

* Fix mistake in test case

* Include missing dependencies

* Remove trailing whitespace

* Dependencies are simply references, not necessarily calls

* More trailing whitespace

* Newline at end of file

* Fix spacing in docstring

Co-authored-by: Siyuan Feng <Hzfengsy@sjtu.edu.cn>

---------

Co-authored-by: Siyuan Feng <Hzfengsy@sjtu.edu.cn>

yelite committed:
[Unity][BYOC] Add batch matmul support to Relax CUTLASS BYOC (#14166)

* Add batch matmul support to Relax CUTLASS BYOC

* Allow more dtypes

* Fix tests

* Revert how to get batch attr

SiriusNEO committed:
[Unity][Op] Full support of Relax op `power` (#14171)

This PR provides a full support of `R.power` including op registering, legalization, overloading `__power__` for Expr and torch fx frontend.

Commit:
Merge remote-tracking branch 'tvm/unity' into relax

driazati reviewed (commented):
What should we do about the CI failures? Just rebase on top of #40?

jwfromm commented:
Closed in favor of #44 which includes a recent rebase on top of `tvm/main`
